src/server/controllers/changePassword.ts

```diff
@@ -0,0 +1,131 @@
+import { Request } from 'express';
+import { Routes } from '@/shared/model/Routes';
+import deepmerge from 'deepmerge';
+import { handleAsyncErrors } from '@/server/lib/expressWrappers';
+import { getBrowserNameFromUserAgent } from '@/server/lib/getBrowserName';
+import { logger } from '@/server/lib/logger';
+import { renderer } from '@/server/lib/renderer';
+import { ResponseWithRequestState } from '@/server/models/Express';
+import {
+  validate as validateToken,
+  change as changePassword,
+} from '@/server/lib/idapi/changePassword';
+import { FieldError } from '@/shared/model/ClientState';
+import { ChangePasswordErrors } from '@/shared/model/Errors';
+import { setIDAPICookies } from '@/server/lib/cookies';
+import { trackMetric } from '@/server/lib/trackMetric';
+import { Metrics } from '@/server/models/Metrics';
+
+const validatePasswordField = (password: string): Array<FieldError> => {
+  const errors: Array<FieldError> = [];
+
+  if (!password) {
+    errors.push({
+      field: 'password',
+      message: ChangePasswordErrors.PASSWORD_BLANK,
+    });
+  } else if (password.length < 8 || password.length > 72) {
+    errors.push({
+      field: 'password',
+      message: ChangePasswordErrors.PASSWORD_LENGTH,
+    });
+  }
+
+  return errors;
+};
+
+export const checkResetPasswordTokenController = (
+  setPasswordPagePath: string,
+  setPasswordPageTitle: string,
+  resendEmailPagePath: string,
+  resendEmailPageTitle: string,
+) =>
+  handleAsyncErrors(async (req: Request, res: ResponseWithRequestState) => {
+    let state = res.locals;
+    const { token } = req.params;
+
+    try {
+      state = deepmerge(state, {
+        pageData: {
+          browserName: getBrowserNameFromUserAgent(req.header('User-Agent')),
+          email: await validateToken(token, req.ip),
+        },
+      });
+
+      const html = renderer(`${setPasswordPagePath}/${token}`, {
+        requestState: state,
+        pageTitle: setPasswordPageTitle,
+      });
+      return res.type('html').send(html);
+    } catch (error) {
+      logger.error(`${req.method} ${req.originalUrl}  Error`, error);
+
+      return res.type('html').send(
+        renderer(`${resendEmailPagePath}${Routes.RESEND}`, {
+          requestState: state,
+          pageTitle: resendEmailPageTitle,
+        }),
+      );
+    }
+  });
+
+export const setPasswordTokenController = (
```

Comment:
Would it make sense to have two controllers, one for setting, one for changing? Or maybe we should change the name of this one from `changePassword` to...? Just a thought

Reply:
Yeah this is a naming problem again, we use "set" and "change" interchangeably as in our case they have exactly the same logic.
Personally I feel that "set" makes more sense than "change" as whether your creating (set) a password for the first time or resetting your password where you "set" a new password.
So maybe the codebase should be renamed throughout?
It's difficult though. Naming things is hard.

Reply:
'set' defo seems more generic, which might be beneficial here. Defo one for a later PR - I don't want to hold this one up